### PR TITLE
[expo-notifications] [ENG-10251] background notifications in foreground

### LINKED ISSUE #24024: [expo-notifications] Background Notifications Not Received When App Is In Foreground
### Minimal reproducible example

https://github.com/warpcast/background-push-demo

### Summary

When sending a background push notification (i.e. with `content-available: 1`) when the app is in the foreground, I would expect at least one event handler (e.g. `setNotificationHandler#handleNotification`, `addNotificationReceivedListener`) to be triggered, but it is not.

## Steps to Reproduce in Demo App
1. In `app.json` and `Info.plist`, replace the bundle identifier (`com.merkle-manufactory.background-push-demo`) with a unique value.
2. Run `yarn install` to install dependencies.
3. Plug in a physical iPhone via USB.
4. Run `yarn ios`, and choose the physical device to build the app.
5. Visit [App Store Connect](https://appstoreconnect.apple.com/apps) and create a new iOS app with the new bundle ID. Be sure to enable the Push Notification capability.
6. Run `yarn ios` to rebuild the app.
7. Enable push notifications via the app's UI.
8. Long-press the device token and copy to clipboard.
9. Visit the [Apple Push Notification Console](https://icloud.developer.apple.com/dashboard/notifications), and navigate to the new push notification form.
10. Paste the copied device token, and with the app **foregrounded**, send a normal **alert** notification.
11. Observe `setNotificationHandler#handleNotification`, `addNotificationReceivedListener`, and `setNotificationHandler#handleSuccess` in the console logs, indicating that the app is receiving the notification.
12. Start a new test push form in the Apple Push Notification Console. Paste the device token, and with the app **backgrounded**, send a **background** notification (i.e. change the `apns-push-type` to `background`, which will add `content-available: 1` to the payload).
13.  Observe `backgroundNotificationTask` in the console logs, indicating that the app is receiving the notification.
14. Return once more to the Apple Push Notification Console to create a new push notification. Send a **background** notification to the device, but this time with the app in the **foreground**.
15. Observe that there are no logs from `setNotificationHandler#handleNotification`, `addNotificationReceivedListener`, `addNotificationResponseReceivedListener`, or `backgroundNotificationTask`. **Expected behavior is that one or more of the previously mentioned event handlers would be triggered**

### Environment

```
 expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 13.5
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 16.20.2 - ~/.volta/tools/image/node/16.20.2/bin/node
      Yarn: 1.22.19 - ~/.volta/tools/image/yarn/1.22.19/bin/yarn
      npm: 8.19.4 - ~/.volta/tools/image/node/16.20.2/bin/npm
      Watchman: 2023.07.24.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.11.3 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
    IDEs:
      Android Studio: 2022.1 AI-221.6008.13.2211.9477386
      Xcode: 14.3.1/14E300c - /usr/bin/xcodebuild
    npmPackages:
      expo: ~49.0.7 => 49.0.7
      react: 18.2.0 => 18.2.0
      react-native: 0.72.3 => 0.72.3
    Expo Workflow: managed
```

### PR DESCRIPTION
**Pull request by kadikraman**
# Why

Background notifications are not sent to the background task when the app is in the foreground.

There isn't a technical limitation for filtering out the background notifications when the app is in the foreground, as the tasks are delivered using [Expo TaskManager](https://docs.expo.dev/versions/latest/sdk/task-manager/) under the hood, which is fine to run in the foreground.

Resolves https://github.com/expo/expo/issues/24024

# How

Execute background notification callback even when the app is in foreground

They were explicitly filtered out on [iOS](https://github.com/expo/expo/blob/main/packages/expo-notifications/ios/EXNotifications/Notifications/Background/EXBackgroundRemoteNotificationConsumer.m#L31) and [Android](https://github.com/expo/expo/blob/main/packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/background/BackgroundRemoteNotificationTaskConsumer.java#L64) so it's just a case of removing this filter.

# Test Plan

Testing locally on a physical device.

- [x] iOS
- [x] Android

Testing details in this comment: https://github.com/expo/expo/pull/24684#issuecomment-1742685316

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by alanjhughes** [comment]
@kadikraman - do we know why this check was here in the first place? 🤔

**Review by christopherwalter — APPROVED** [review]
LGTM. How did local testing go? Anything to indicate why we might've written it this way to begin with?

**Comment by kadikraman** [comment]
@alanjhughes we think it was just an assumption - _background_ notification, _background_ task, why would you need to run it when the app is in the foreground. Haven't found anything is code to indicate otherwise.

@christopherwalter tested on device on iOS and it works as expected. Still need to confirm on Android.

**Comment by kadikraman** [comment]
@christopherwalter @alanjhughes tested locally on both iOS and Android now and I'm confident this works as expected: by not filtering out the background requests when the app is in the foreground, the background task just gets run. So in my case, the "IT WORKS!!!!!" gets printed out which is exactly what I'd expect.

This is how I tested:

Created an expo app with `npx create-expo-app` and installed `npx expo install expo-notification expo-task-manager`

Added the following in the main entry file (outside a React component):

```js
import * as Notifications from 'expo-notifications';
import * as TaskManager from 'expo-task-manager';

TaskManager.defineTask("BACKGROUND_NOTIFICATION_TASK", (_data) => {
  console.log("IT WORKS!!!!!");
});
Notifications.registerTaskAsync("BACKGROUND_NOTIFICATION_TASK");
```

## iOS
Requested push permission with `await Notifications.requestPermissionsAsync();`. Then tested via the [Apple Push Notification Console ](https://icloud.developer.apple.com/dashboard/notifications) (note the `Type = Background` which makes this a background notification).

Screenshot 2023-10-02 at 09 57 20


## Android
There is no UI available for Android to test data-only notifications, so I had to set up FCM for the project, enable the legacy cloud messaging API and use the server key send the massage as an API request:

```curl
curl --request POST \
  --url https://fcm.googleapis.com/fcm/send \
  --header 'Authorization: key=<api-key>' \
  --header 'Content-Type: application/json' \
  --header 'User-Agent: insomnia/8.1.0' \
  --data '{
	"data": {
		"hello": "world"
	},
	"registration_ids": [
		<device-token>
	]
}'
```

Screenshot 2023-10-02 at 10 14 19

File: docs/pages/versions/v49.0.0/sdk/notifications.mdx [hunk 1]
```diff
@@ -417,24 +417,28 @@ or interacting with it via [notification categories](#managing-notification-cate
...
+| true                               |          Killed          | `NotificationResponseReceivedListener`                                  |
 
 In the table above, whenever `NotificationResponseReceivedListener` is triggered, the same would apply to the `useLastNotificationResponse` hook.
 
-### Background events
+### Background notifications
 
 > **warning** Background event listeners are not supported in Expo Go.
 
+A background notification (iOS) or a data-only notifiation (Android) is a remote notification that does not display an alert, play a sound, or add a badge to your app's icon. The purpose of a background notification is to give you a way to wake up your app to trigger an app data refresh in the background.
```
Comment (amandeepmittal) [inline docs/pages/versions/v49.0.0/sdk/notifications.mdx:435]:
```suggestion
A background notification (iOS) or a data-only notification (Android) is a remote notification that does not display an alert, play a sound, or add a badge to your app's icon. The purpose of a background notification is to provide a way to wake up your app to trigger an app data refresh in the background.
```

In the suggestion:
- Fixed typo
- a bit more concise

**Review by amandeepmittal — COMMENTED** [review]
Can you please update the [`versions/unversioned/sdk/notifications.mdx`](https://github.com/expo/expo/blob/%40kadikraman/remote-notifications/docs/pages/versions/unversioned/sdk/notifications.mdx) so that the changes are not lost when we cut the docs for the next SDK? When we release a new SDK, we copy the unversioned directory and rename it to the new version. more context on this in [`docs/readme`](https://github.com/expo/expo/blob/main/docs/README.md#update-latest-version-of-docs).


Also, left a small suggestion

**Review by Simek — COMMENTED** [review]
Since you have modified the code comment in package source we also need to regenerate the API data for that package. The guide how to do that can be found here:
* https://github.com/expo/expo/tree/main/docs#update-api-reference-docs

**Review by alanjhughes — APPROVED** [review]

**Comment by kadikraman** [comment]
The changes in this PR don't have any TypeScript changes so I reverted the API docs comment.
The iOS and web tests are unrelated to this PR.

**Pull request merged**